src/buildercore/threadbare/execute.py

```diff
@@ -0,0 +1,225 @@
+        queue.put({"name": name, "result": result})
+    except BaseException as unhandled_exception:
+        # kept for debugging
+        # import traceback
+        # traceback.print_exc()
+
+        # "Note that exit handlers and finally clauses, etc., will not be executed."
+        # - https://docs.python.org/2/library/multiprocessing.html#multiprocessing.Process.terminate
+        queue.put({"name": name, "result": unhandled_exception})
+
+
+def process_status(running_p):
+    # https://docs.python.org/2/library/multiprocessing.html#process-and-exceptions
+    result = {
+        "pid": running_p.pid,
+        "name": running_p.name,
+        "exitcode": running_p.exitcode,
+        "alive": running_p.is_alive(),
+        "killed": False,
+        "kill-signal": None,
+    }
+    if running_p.exitcode is not None and running_p.exitcode < 0:
+        result["killed"] = True
+        result["kill-signal"] = -running_p.exitcode
+    return result
+
+
+def _parallel_execution(env, func, param_key, param_values, return_process_pool=False):
+    "executes the given function in parallel to main process. blocks until processes are complete"
+    results_q = Queue()
+    kwargs = {
+        #'env': ..., # each process will get a new state dictionary
+        "worker_func": func,
+        #'name': ..., # a name is assigned on process start
+        "queue": results_q,
+    }
+    pool_size = getattr(func, "pool_size", None)
+    pool_size = pool_size if pool_size is not None else 1
+    pool_values = param_values or range(0, pool_size)
+
+    pool = []
+    for idx, nth_val in enumerate(pool_values):
+        kwargs["name"] = "process--" + str(idx + 1)  # process--1, process--2
+        new_env = {} if not env else copy.deepcopy(env)
+
+        # ssh clients are not shared between processes
+        if "ssh_client" in new_env:
+            del new_env["ssh_client"]
+
+        if param_key:
+            new_env[param_key] = nth_val
+
+        new_env["parallel"] = True
+        # https://github.com/mathiasertl/fabric/blob/master/fabric/tasks.py#L223-L227
+        # new_env['linewise'] = True # not set until needed
+
+        kwargs["env"] = new_env
+        p = Process(
+            name=kwargs["name"],
+            target=_parallel_execution_worker_wrapper,
+            kwargs=kwargs,
+        )
+        p.start()
+        pool.append(p)
+
+    if return_process_pool:
+        # don't poll for results, don't wait to finish, just return the list of running processes
+        return results_q, pool
+
+    result_map = {}  # {process-name: process-results, ...}
+
+    # poll the processes until all are complete
+    # remove process from pool when it is complete
+    while len(pool) > 0:
+        for idx, running_p in enumerate(pool):
+            result = process_status(running_p)
+            if not result["alive"]:
+                result_map[result["name"]] = result
+                del pool[idx]
+        time.sleep(0.1)
```

Comment:
magic sleep? what does it do?

Reply:
ah, polling interval; usually I ask for a timeout here but this could be arbitrarily long, deploys that take tens of minutes due to backup being made, etc.

Reply:
yes, it introduces the slightest of delays so that we're not manically polling the process every microsecond. I'm happy to implement a longer poll duration if you're expecting a long running process. Failed processes would obviously take longer to reveal themselves.

timeouts are also available in `local`, `remote` and `remote_sudo`. `local` timeouts use Python's `Timer` to kill the process. Remote timeouts rely on pssh's implementation.

I'll add a note as to why we're sleeping here